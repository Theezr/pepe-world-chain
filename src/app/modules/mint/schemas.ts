import { LENGTH_COLLECTION_ID, MAX_LENGTH_MODULE_NAME, MIN_LENGTH_MODULE_NAME } from './constants';

export const mintSchema = {
	$id: 'BaseMintCommand',
	type: 'object',
	properties: {
		tokenID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		amount: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		recipient: {
			dataType: 'bytes',
			fieldNumber: 3,
			format: 'klayr32',
		},
	},
	required: ['tokenID', 'amount', 'recipient'],
};
1;
export const createNftSchema = {
	$id: 'CreateNftCommand',
	type: 'object',
	properties: {
		recipient: {
			dataType: 'bytes',
			fieldNumber: 3,
			format: 'klayr32',
		},
	},
	required: ['recipient'],
};

export const mintNftParamsSchema = {
	$id: '/klayr/nftTransferParams',
	type: 'object',
	required: ['address', 'collectionID', 'attributesArray'],
	properties: {
		address: {
			dataType: 'bytes',
			format: 'klayr32',
			fieldNumber: 1,
		},
		collectionID: {
			dataType: 'bytes',
			minLength: LENGTH_COLLECTION_ID,
			maxLength: LENGTH_COLLECTION_ID,
			fieldNumber: 2,
		},
		attributesArray: {
			type: 'array',
			fieldNumber: 3,
			items: {
				type: 'object',
				required: ['module', 'attributes'],
				properties: {
					module: {
						dataType: 'string',
						minLength: MIN_LENGTH_MODULE_NAME,
						maxLength: MAX_LENGTH_MODULE_NAME,
						pattern: '^[a-zA-Z0-9]*$',
						fieldNumber: 1,
					},
					attributes: {
						dataType: 'bytes',
						fieldNumber: 2,
					},
				},
			},
		},
	},
};

export const upgradeNftParamsSchema = {
	$id: '/klayr/upgradeNftParams',
	type: 'object',
	required: ['nftID', 'module', 'attributes'],
	properties: {
		nftID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		module: {
			dataType: 'string',
			minLength: MIN_LENGTH_MODULE_NAME,
			maxLength: MAX_LENGTH_MODULE_NAME,
			pattern: '^[a-zA-Z0-9]*$',
			fieldNumber: 2,
		},
		attributes: {
			dataType: 'bytes',
			fieldNumber: 3,
		},
	},
};
