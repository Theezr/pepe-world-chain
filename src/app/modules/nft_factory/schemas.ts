export const createFirstPepeSchema = {
	$id: 'CreateFirstPepeCommand',
	type: 'object',
	properties: {
		recipient: {
			dataType: 'bytes',
			fieldNumber: 1,
			format: 'klayr32',
		},
	},
	required: ['recipient'],
};
