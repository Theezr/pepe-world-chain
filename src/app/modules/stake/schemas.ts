export const stakePepeSchema = {
	$id: 'StakePepeCommand',
	type: 'object',
	properties: {
		nftID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
	},
	required: ['nftID'],
};

export const unstakePepeSchema = {
	$id: 'UnstakePepeCommand',
	type: 'object',
	properties: {
		nftID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
	},
	required: ['nftID'],
};

export const claimRewardsSchema = {
	$id: 'ClaimRewardsCommand',
	type: 'object',
	properties: {
		nftID: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
	},
	required: ['nftID'],
};

export const createFirstSchema = {
	$id: 'CreateFirstPepeCommand',
	type: 'object',
	properties: {
		recipient: {
			dataType: 'bytes',
			fieldNumber: 1,
			format: 'klayr32',
		},
	},
	required: ['recipient'],
};
