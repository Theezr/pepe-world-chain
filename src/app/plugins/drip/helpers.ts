export function decodeAttributes(hexString: string) {
	if (!hexString) return {};

	const buffer = Buffer.from(hexString, 'hex');
	const jsonString = buffer.toString('utf8');
	const attributes = JSON.parse(jsonString);
	return attributes;
}

export function wait(ms: number) {
	return new Promise(resolve => setTimeout(resolve, ms));
}
