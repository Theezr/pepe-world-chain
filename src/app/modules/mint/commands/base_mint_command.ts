/* eslint-disable class-methods-use-this */
import { TokenMethod } from 'klayr-framework/dist-node/modules/token';
import { Modules, StateMachine } from 'klayr-sdk';
import { mintSchema } from '../schemas';
import { TokenID } from 'klayr-framework/dist-node/modules/token/types';

interface MintParams {
	tokenID: TokenID;
	amount: bigint;
	recipient: string;
}

export class BaseMintCommand extends Modules.BaseCommand {
	private _tokenMethod!: TokenMethod;

	public schema = mintSchema;

	public addDependencies(args: { tokenMethod: TokenMethod }) {
		this._tokenMethod = args.tokenMethod;
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async verify(
		context: StateMachine.CommandVerifyContext<MintParams>,
	): Promise<StateMachine.VerificationResult> {
		const { amount } = context.params;

		console.log({ amount });

		return { status: StateMachine.VerifyStatus.OK };
	}

	public async execute(context: StateMachine.CommandExecuteContext<MintParams>): Promise<void> {
		const { recipient, amount, tokenID } = context.params;
		console.log('executing mint command');
		console.log({ recipient, amount, tokenID });

		await this._tokenMethod.initializeToken(context.getMethodContext(), Buffer.from(tokenID));
		await this._tokenMethod.mint(
			context.getMethodContext(),
			Buffer.from(recipient),
			Buffer.from(tokenID),
			BigInt(amount),
		);
	}
}
