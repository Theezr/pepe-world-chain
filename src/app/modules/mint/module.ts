/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */
import { NFTMethod } from 'klayr-framework/dist-node/modules/nft/method';
import { TokenMethod } from 'klayr-framework/dist-node/modules/token';
import { Modules } from 'klayr-sdk';
import { BaseMintCommand } from './commands/base_mint_command';
import { CreateNftCommand } from './commands/create_nft_command';
import { UpgradeNftCommand } from './commands/upgrade_nft_command';
import { MintEndpoint } from './endpoint';
import { MintMethod } from './method';

export class MintModule extends Modules.BaseModule {
	public endpoint = new MintEndpoint(this.stores, this.offchainStores);
	public method = new MintMethod(this.stores, this.events);
	public _baseMintCommand = new BaseMintCommand(this.stores, this.events);
	public _createNftCommand = new CreateNftCommand(this.stores, this.events);
	public _upgradeNftCommand = new UpgradeNftCommand(this.stores, this.events);

	public commands = [this._baseMintCommand, this._createNftCommand, this._upgradeNftCommand];

	public _tokenMethod!: TokenMethod;
	public _nftMethod!: NFTMethod;

	public constructor() {
		super();
		// registeration of stores and events
	}

	public addDependencies(args: { tokenMethod: TokenMethod; nftMethod: NFTMethod }): void {
		this._tokenMethod = args.tokenMethod;
		this._nftMethod = args.nftMethod;

		this._baseMintCommand.addDependencies({
			tokenMethod: this._tokenMethod,
		});

		this._createNftCommand.addDependencies({
			nftMethod: this._nftMethod,
		});

		this._upgradeNftCommand.addDependencies({
			nftMethod: this._nftMethod,
		});
	}

	public metadata(): Modules.ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [],
			assets: [],
		};
	}

	// Lifecycle hooks
	public async init(_args: Modules.ModuleInitArgs): Promise<void> {
		// initialize this module when starting a node
	}

	// public async insertAssets(_context: StateMachine.InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: StateMachine.BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(
	// 	_context: StateMachine.TransactionVerifyContext,
	// ): Promise<StateMachine.VerificationResult> {
	// 	// verify transaction will be called multiple times in the transaction pool
	// 	return { status: StateMachine.VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(
	// 	context: StateMachine.TransactionExecuteContext,
	// ): Promise<void> {
	// 	const test = await context
	// 		.getMethodContext()
	// 		.contextStore.set('CONTEXT_STORE_KEY_AVAILABLE_FEE', BigInt(100000000000000));
	// 	console.log('test', test);
	// }

	// public async afterCommandExecute(_context: StateMachine.TransactionExecuteContext): Promise<void> {

	// }
	// public async initGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async finalizeGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async beforeTransactionsExecute(_context: StateMachine.BlockExecuteContext): Promise<void> {

	// }

	// public async afterTransactionsExecute(_context: StateMachine.BlockAfterExecuteContext): Promise<void> {

	// }
}
