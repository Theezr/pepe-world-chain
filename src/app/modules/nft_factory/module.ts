/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */
import { Modules } from 'klayr-sdk';
import { CreateFirstPepeCommand } from './commands/create_first_pepe_command';
import { NftFactoryEndpoint } from './endpoint';
import { NftFactoryMethod } from './method';
import { NFTMethod } from 'klayr-framework/dist-node/modules/nft';

export class NftFactoryModule extends Modules.BaseModule {
	public endpoint = new NftFactoryEndpoint(this.stores, this.offchainStores);
	public method = new NftFactoryMethod(this.stores, this.events);
	public _createFirstPepeCommand = new CreateFirstPepeCommand(this.stores, this.events);
	public commands = [this._createFirstPepeCommand];

	public _nftMethod!: NFTMethod;

	public constructor() {
		super();
		// registeration of stores and events
	}

	public addDependencies(args: { nftMethod: NFTMethod }): void {
		this._nftMethod = args.nftMethod;

		this._createFirstPepeCommand.addDependencies({
			nftMethod: this._nftMethod,
		});
	}

	public metadata(): Modules.ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [],
			assets: [],
		};
	}

	// Lifecycle hooks
	// public async init(_args: Modules.ModuleInitArgs): Promise<void> {
	// 	// initialize this module when starting a node
	// }

	// public async insertAssets(_context: StateMachine.InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: StateMachine.BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(_context: StateMachine.TransactionVerifyContext): Promise<StateMachine.VerificationResult> {
	// verify transaction will be called multiple times in the transaction pool
	// return { status: StateMachine.VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(_context: StateMachine.TransactionExecuteContext): Promise<void> {
	// }

	// public async afterCommandExecute(_context: StateMachine.TransactionExecuteContext): Promise<void> {

	// }
	// public async initGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async finalizeGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async beforeTransactionsExecute(_context: StateMachine.BlockExecuteContext): Promise<void> {

	// }

	// public async afterTransactionsExecute(_context: StateMachine.BlockAfterExecuteContext): Promise<void> {

	// }
}
