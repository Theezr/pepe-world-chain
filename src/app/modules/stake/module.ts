/* eslint-disable class-methods-use-this */
/* eslint-disable @typescript-eslint/member-ordering */
import { NFTMethod } from 'klayr-framework/dist-node/modules/nft';
import { Modules } from 'klayr-sdk';
import { ClaimRewardsCommand } from './commands/claim_rewards_command';
import { CreateFirstBusinessCommand } from './commands/create_first_business_command';
import { CreateFirstPepeCommand } from './commands/create_first_pepe_command';
import { LevelWorkerCommand } from './commands/level_worker_command';
import { StakePepeCommand } from './commands/stake_pepe_command';
import { UnstakePepeCommand } from './commands/unstake_pepe_command';
import { UpgradeBusinessCommand } from './commands/upgrade_business_command';
import { StakeEndpoint } from './endpoint';
import { StakeMethod } from './method';
import { BusinessStore } from './stores/businessStore';
import { StakeTimeStore } from './stores/stakeTime';
import { WorkerStore } from './stores/workerStore';
import { WorkerStakedStore } from './stores/workerStakedStore';

export class StakeModule extends Modules.BaseModule {
	public endpoint = new StakeEndpoint(this.stores, this.offchainStores);
	public method = new StakeMethod(this.stores, this.events);

	public _stakePepeCommand = new StakePepeCommand(this.stores, this.events);
	public _unstakePepeCommand = new UnstakePepeCommand(this.stores, this.events);
	public _claimRewardsCommand = new ClaimRewardsCommand(this.stores, this.events);
	public _createFirstPepeCommand = new CreateFirstPepeCommand(this.stores, this.events);
	public _createFirstBusinessCommand = new CreateFirstBusinessCommand(this.stores, this.events);
	public _upgradeBusinessCommand = new UpgradeBusinessCommand(this.stores, this.events);
	public _levelWorkerCommand = new LevelWorkerCommand(this.stores, this.events);

	public commands = [
		this._stakePepeCommand,
		this._claimRewardsCommand,
		this._unstakePepeCommand,
		this._createFirstPepeCommand,
		this._createFirstBusinessCommand,
		this._upgradeBusinessCommand,
		this._levelWorkerCommand,
	];

	public _nftMethod!: NFTMethod;
	public _tokenMethod!: Modules.Token.TokenMethod;

	public constructor() {
		super();
		// registeration of stores and events
		this.stores.register(StakeTimeStore, new StakeTimeStore(this.name, 0));
		this.stores.register(WorkerStore, new WorkerStore(this.name, 1));
		this.stores.register(BusinessStore, new BusinessStore(this.name, 2));
		this.stores.register(WorkerStakedStore, new WorkerStakedStore(this.name, 3));
	}

	public addDependencies(args: {
		tokenMethod: Modules.Token.TokenMethod;
		nftMethod: NFTMethod;
	}): void {
		this._nftMethod = args.nftMethod;
		this._tokenMethod = args.tokenMethod;

		this._stakePepeCommand.addDependencies({
			nftMethod: this._nftMethod,
		});

		this._unstakePepeCommand.addDependencies({
			method: this.method,
			nftMethod: this._nftMethod,
		});

		this._claimRewardsCommand.addDependencies({
			method: this.method,
			nftMethod: this._nftMethod,
		});

		this._levelWorkerCommand.addDependencies({
			method: this.method,
			nftMethod: this._nftMethod,
		});

		this._createFirstPepeCommand.addDependencies({
			method: this.method,
			nftMethod: this._nftMethod,
		});

		this._createFirstBusinessCommand.addDependencies({
			method: this.method,
			nftMethod: this._nftMethod,
		});

		this._upgradeBusinessCommand.addDependencies({
			method: this.method,
			tokenMethod: this._tokenMethod,
			nftMethod: this._nftMethod,
		});

		this.method.addDependencies({
			tokenMethod: this._tokenMethod,
			nftMethod: this._nftMethod,
		});
		this.endpoint.addDependencies(this.method);
	}

	public metadata(): Modules.ModuleMetadata {
		return {
			...this.baseMetadata(),
			endpoints: [
				// {
				// 	name: this.endpoint.getStakeRewards.name,
				// 	request: getStakerRequestSchema,
				// 	response: getStakerResponseSchema,
				// },
			],
			assets: [],
		};
	}

	// Lifecycle hooks
	// public async init(_args: Modules.ModuleInitArgs): Promise<void> {
	// 	// initialize this module when starting a node
	// }

	// public async insertAssets(_context: StateMachine.InsertAssetContext) {
	// 	// initialize block generation, add asset
	// }

	// public async verifyAssets(_context: StateMachine.BlockVerifyContext): Promise<void> {
	// 	// verify block
	// }

	// Lifecycle hooks
	// public async verifyTransaction(_context: StateMachine.TransactionVerifyContext): Promise<StateMachine.VerificationResult> {
	// verify transaction will be called multiple times in the transaction pool
	// return { status: StateMachine.VerifyStatus.OK };
	// }

	// public async beforeCommandExecute(_context: StateMachine.TransactionExecuteContext): Promise<void> {
	// }

	// public async afterCommandExecute(_context: StateMachine.TransactionExecuteContext): Promise<void> {

	// }
	// public async initGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async finalizeGenesisState(_context: StateMachine.GenesisBlockExecuteContext): Promise<void> {

	// }

	// public async beforeTransactionsExecute(_context: StateMachine.BlockExecuteContext): Promise<void> {

	// }

	// public async afterTransactionsExecute(_context: StateMachine.BlockAfterExecuteContext): Promise<void> {

	// }
}
