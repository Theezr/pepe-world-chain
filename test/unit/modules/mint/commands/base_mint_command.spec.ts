import { utils } from '@klayr/cryptography';
import { BaseMintCommand } from '../../../../../src/app/modules/mint/commands/base_mint_command';
import { MintModule } from '../../../../../src/app/modules/mint/module';
import { chain, codec, db, Transaction } from 'klayr-sdk';
import { mintSchema } from '../../../../../src/app/modules/mint/schemas';
import { createCtx, createSampleTransaction } from '../../../../helpers';
import { CommandExecuteContext, VerifyStatus } from 'klayr-framework/dist-node/state_machine';

interface MintParams {
	amount: bigint;
	recipient: string;
}

const mockMint = jest.fn();

describe('MintModule', () => {
	let baseMintCommand: BaseMintCommand;
	let stateStore: any;

	beforeEach(() => {
		const mintModule = new MintModule();
		baseMintCommand = new BaseMintCommand(mintModule.stores, mintModule.events);
		baseMintCommand.addDependencies({
			tokenMethod: { mint: mockMint } as any,
		});

		stateStore = new chain.StateStore(new db.InMemoryDatabase());
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(baseMintCommand.name).toEqual('baseMint');
		});

		it('should have valid schema', () => {
			expect(baseMintCommand.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
			it('should throw when `amount` is too high', async () => {
				const params = codec.encode(mintSchema, {
					amount: BigInt(2e6),
					recipient: utils.getRandomBytes(20),
				});
				const transaction = new Transaction(createSampleTransaction(params, BaseMintCommand.name));
				const ctx = createCtx<MintParams>(stateStore, transaction, mintSchema, 'verify');

				const result = await baseMintCommand.verify(ctx);
				expect(result.status).toBe(VerifyStatus.OK);
			});
			it.todo('should be ok for valid schema');
		});
	});

	describe('execute', () => {
		describe('valid cases', () => {
			it('mint function should have been called` ', async () => {
				const params = codec.encode(mintSchema, {
					amount: BigInt(2e6),
					recipient: utils.getRandomBytes(20),
				});
				const transaction = new Transaction(createSampleTransaction(params, BaseMintCommand.name));
				const ctx = createCtx<MintParams>(stateStore, transaction, mintSchema, 'execute');

				await expect(
					baseMintCommand.execute(ctx as CommandExecuteContext<MintParams>),
				).resolves.toBeUndefined();
			});

			it.todo('should update the state store');
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
