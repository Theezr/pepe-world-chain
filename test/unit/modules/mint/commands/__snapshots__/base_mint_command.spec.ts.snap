// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MintModule constructor should have valid schema 1`] = `
{
  "$id": "BaseMintCommand",
  "properties": {
    "amount": {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
    "recipient": {
      "dataType": "bytes",
      "fieldNumber": 3,
      "format": "klayr32",
    },
  },
  "required": [
    "amount",
    "recipient",
  ],
  "type": "object",
}
`;
